### PR TITLE
added specifications and implemented them in GetAll for projects

### PR DESCRIPTION
**Pull request by Z1rael88**
Could you please check specifications implementation in GetAllProjects

### CODE REVIEW

File: Application/Services/ProjectService.cs [hunk 1]
```diff
@@ -27,7 +28,7 @@ public async Task<IEnumerable<ProjectResponse>> GetAllAsync(ProjectSearchParams
             return cachedProjects;
         }
 
-        var projects = await projectRepository.GetAllAsync(projectSearchParams);
+        var projects = await repository.GetAllAsync(new ProjectWithTasksSpecifications(projectSearchParams));
```
Comment (ilya-valentsiukevich) [inline Application/Services/ProjectService.cs]: use more specific name, projectRepository for example

File: Application/Services/TaskService.cs [hunk 2]
```diff
@@ -57,13 +57,13 @@ public async Task<TaskResponse> CreateAsync(CreateTaskRequest createTaskRequest)
     {
         logger.LogInformation("Started creating task from Service Layer");
         await createTaskValidator.ValidateAndThrowAsync(createTaskRequest);
-        if (!await projectRepository.IsProjectExistsAsync(createTaskRequest.ProjectId))
+        if (!await repository.IsEntityExistsAsync(createTaskRequest.ProjectId))
```
Comment (ilya-valentsiukevich) [inline Application/Services/TaskService.cs]: it doesn't work, you're trying to get a project using task repo, you need 2 different repositories for this (task and project)

File: Domain/Extensions/ExpressionExtension.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+using System.Linq.Expressions;
+
+namespace Domain.Extensions;
+
+public static class ExpressionExtension
```
Comment (ilya-valentsiukevich) [inline Domain/Extensions/ExpressionExtension.cs]: I don't agree, that this file should be on Domain side

File: Domain/Interfaces/IApplicationDbContext.cs [hunk 4]
```diff
@@ -8,4 +8,5 @@ public interface IApplicationDbContext
     DbSet<ProjectModel> Projects { get; set; }
     DbSet<TaskModel> Tasks { get; set; }
     Task<int> SaveChangesAsync(CancellationToken cancellationToken = default);
+    DbSet<TEntity> GetTable<TEntity>() where TEntity : BaseModel;
```
Comment (ilya-valentsiukevich) [inline Domain/Interfaces/IApplicationDbContext.cs]: Use GetDbSet name instead of GetTable :)

File: Infrastructure/Repositories/Repository.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        return await queryableResultWithIncludes
+            .Where(specification.Criteria)
+            .ToListAsync();
+    }
+
+    public async Task<bool> IsEntityExistsAsync(Guid id)
+    {
+        logger.LogInformation($"Started checking if there is {typeof(T).Name} with {id} in database");
+        var response = await dbContext.GetTable<T>().AnyAsync(p => p.Id == id);
+        logger.LogInformation($"Successfully found {typeof(T).Name} with {id} in database");
```
Comment (ilya-valentsiukevich) [inline Infrastructure/Repositories/Repository.cs]: Incorrect log, maybe we don't have an entity with that id, but you log that we have in any case

File: Infrastructure/Repositories/Repository.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+public class Repository<T>(IApplicationDbContext dbContext, ILogger<Repository<T>> logger)
+    : IRepository<T> where T : BaseModel
+{
+    public async Task<T> CreateAsync(T entity)
+    {
+        logger.LogInformation($"Started creating {nameof(T)} from database");
+        var newEntity = dbContext.GetTable<T>().Add(entity);
+        await dbContext.SaveChangesAsync();
+        logger.LogInformation($"Successfully created {typeof(T).Name} from database");
```
Comment (ilya-valentsiukevich) [inline Infrastructure/Repositories/Repository.cs]: use nameof(T) instead of typeof

File: Application/Specification/BaseSpecification.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+using Domain.Models;
+
+namespace Application.Specification
+{
+    public abstract class BaseSpecification<T>(Expression<Func<T, bool>> criteria) : ISpecification<T>
+        where T : BaseModel
+    {
+        public Expression<Func<T, bool>> Criteria { get; } = criteria;
+        public List<Expression<Func<T, object>>> Includes { get; } = new List<Expression<Func<T, object>>>();
+        public List<string> IncludeStrings { get; } = new List<string>();
```
Comment (ilya-valentsiukevich) [inline Application/Specification/BaseSpecification.cs]: remove unused IncludeStrings

File: Application/Specification/BaseSpecification.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+        protected virtual void AddInclude(Expression<Func<T, object>> includeExpression)
+        {
+            Includes.Add(includeExpression);
+        }
+
+        protected virtual void AddInclude(string includeString)
```
Comment (ilya-valentsiukevich) [inline Application/Specification/BaseSpecification.cs]: remove unused AddInclude method

File: Application/Services/TaskService.cs (hunk 2 above)
Comment (Z1rael88) [reply Application/Services/TaskService.cs]: So i need to get my repositories back>????)

File: Domain/Extensions/ExpressionExtension.cs (hunk 3 above)
Comment (Z1rael88) [reply Domain/Extensions/ExpressionExtension.cs]: Application?

File: Application/Services/TaskService.cs (hunk 2 above)
Comment (ilya-valentsiukevich) [reply Application/Services/TaskService.cs]: of course not, you can inject two generic repo with different T

Comment (Z1rael88) [reply Application/Services/TaskService.cs]: oh okay!

File: Domain/Extensions/ExpressionExtension.cs (hunk 3 above)
Comment (ilya-valentsiukevich) [reply Domain/Extensions/ExpressionExtension.cs]: for example yes

**Pull request merged**